### PR TITLE
introduce install ID, capture anonymous usage stats. #658

### PR DESCRIPTION
**Pull request by sriv**
- generate a unique UUID per install. #653 
- track and send anonymous events to GA, #658

### CODE REVIEW

File: skel/gauge.properties [hunk 1]
```diff
@@ -20,3 +20,9 @@ runner_request_timeout = 30000
 
 # Allow Gauge and its plugin updates to be notified.
 check_updates = true
+
+# Allow Gauge to collect anonymous usage statistics
+gauge_analytics_enabled = true
```
Comment (kashishm) [inline skel/gauge.properties]: The value should be `false` here. The analytics should be turned off by default.

File: filter/specItemFilter.go [hunk 2]
```diff
@@ -77,25 +76,15 @@ func (filter *ScenarioFilterBasedOnTags) replaceSpecialChar() {
 }
 
 func (filter *ScenarioFilterBasedOnTags) formatAndEvaluateExpression(tagsMap map[string]bool, isTagQualified func(tagsMap map[string]bool, tagName string) bool) (bool, error) {
-	_, tags := filter.getOperatorsAndOperands()
-	expToBeEvaluated := filter.tagExpression
-	sort.Sort(ByLength(tags))
+	tagExpressionParts, tags := filter.parseTagExpression()
```
Comment (kashishm) [inline filter/specItemFilter.go]: Why are these changes showing up in this PR. These changes are already merged in #671

File: config/configuration.go [hunk 3]
```diff
@@ -121,6 +136,22 @@ func SetProjectRoot(args []string) error {
...
 }
 
+// UniqueID gets the unique installation ID.
+func UniqueID() string {
+	configDir, err := common.GetConfigurationDir()
+	if err != nil {
+		APILog.Warning("Unable to read config dir, %s", err)
+		return ""
+	}
+	idFile := filepath.Join(configDir, ".gauge_id")
```
Comment (kashishm) [inline config/configuration.go]: The file is already in `.gauge` folder. So, the filename can just be `.id` or `id`. Thoughts?

File: config/configuration_test.go [hunk 4]
```diff
@@ -46,27 +42,70 @@ func stub4GetFromConfig(propertyName string) string {
...
+	expected := "foo"
+	idFile := filepath.Join("_testData", ".gauge_id")
+	ioutil.WriteFile(idFile, []byte(expected), common.NewFilePermissions)
+
+	s, err := filepath.Abs("_testData")
+	if err != nil {
+		t.Error(err)
+	}
+
+	os.Setenv("GAUGE_ROOT", s)
```
Comment (kashishm) [inline config/configuration_test.go]: We should be using `GAUGE_HOME`. `GAUGE_ROOT` is just for backward compatibility.

Comment (sriv) [inline config/configuration_test.go]: we use `common.GetConfigurationDir` which returns `gauge_root` or `$home/.gauge/config`

File: track/track.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+func Execution(parallel, tagged, sorted, simpleConsole, verbose bool, parallelExecutionStrategy string) {
+	action := "serial"
+	if parallel {
+		action = "parallel"
+	}
+
+	label := ""
+	if tagged {
+		label = "tagged,"
```
Comment (kashishm) [inline track/track.go]: Why is `,` required at the end. If I am execution via tags(no other flag) then the label would be `tagged,`. Is that intentional? May be we can use array and join it at the end with `,`

File: config/configuration_test.go (hunk 4 above)
Comment (kashishm) [inline config/configuration_test.go]: Ok. I think we should change that also to use `GAUGE_HOME`. But this is out of scope for this PR

File: track/track.go (hunk 5 above)
Comment (sriv) [inline track/track.go]: fair point. sorted in the next commit.
